File: Backend/src/main/java/com/Trabajo_Final_Beltran/service/ChatService.java
package com.Trabajo_Final_Beltran.service;

import com.Trabajo_Final_Beltran.dto.request.EnviarMensajeRequest;
import com.Trabajo_Final_Beltran.dto.response.ChatConversacionResponse;
import com.Trabajo_Final_Beltran.dto.response.ChatMensajeResponse;
import com.Trabajo_Final_Beltran.entity.ChatConversacion;
import com.Trabajo_Final_Beltran.entity.Usuario;
import com.Trabajo_Final_Beltran.enums.EstadoConversacion;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ChatService {

    ChatConversacion obtenerConversacionAutorizada(Long conversacionId, Usuario usuario);

    ChatConversacionResponse crearOReutilizarChat(Long pedidoId);

    ChatConversacionResponse obtenerDetalle(Long conversacionId);

    Page<ChatMensajeResponse> obtenerMensajes(Long conversacionId, Pageable pageable);

    Page<ChatConversacionResponse> listarPorEstablecimiento(EstadoConversacion estado, Pageable pageable);

    ChatMensajeResponse enviarMensaje(Long conversacionId, EnviarMensajeRequest request, Usuario usuario);
    
    ChatMensajeResponse enviarMensaje(Long conversacionId, EnviarMensajeRequest request);
    
    ChatConversacionResponse asignar(Long conversacionId);

    ChatConversacionResponse cerrar(Long conversacionId);

    com.Trabajo_Final_Beltran.dto.response.ChatLecturaResponse marcarMensajesComoLeidos(Long conversacionId, Usuario usuario);
}

File: frontEnd/src/app/core/services/chat.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  ChatConversacionResponse,
  ChatMensajeResponse,
  CrearChatRequest,
  EstadoConversacion
} from '../models/chat.models';

export interface SpringPage<T> {
  content: T[];
  totalPages: number;
  totalElements: number;
  last: boolean;
  first: boolean;
  number: number;
  size: number;
  empty: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class ChatService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.baseUrl}/chats`;

  crearOReutilizarChat(pedidoId: number): Observable<ChatConversacionResponse> {
    const body: CrearChatRequest = { pedidoId };
    return this.http.post<ChatConversacionResponse>(this.apiUrl, body);
  }

  listar(
    estado?: EstadoConversacion,
    page: number = 0,
    size: number = 20
  ): Observable<SpringPage<ChatConversacionResponse>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    if (estado) {
      params = params.set('estado', estado);
    }
    return this.http.get<SpringPage<ChatConversacionResponse>>(this.apiUrl, { params });
  }

  asignar(conversacionId: number): Observable<ChatConversacionResponse> {
    return this.http.patch<ChatConversacionResponse>(
      `${this.apiUrl}/${conversacionId}/asignar`,
      {}
    );
  }

  obtenerDetalle(conversacionId: number): Observable<ChatConversacionResponse> {
    return this.http.get<ChatConversacionResponse>(`${this.apiUrl}/${conversacionId}`);
  }

  obtenerMensajes(
    conversacionId: number,
    page: number = 0,
    size: number = 20
  ): Observable<SpringPage<ChatMensajeResponse>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<SpringPage<ChatMensajeResponse>>(
      `${this.apiUrl}/${conversacionId}/mensajes`,
      { params }
    );
  }

  cerrar(conversacionId: number): Observable<ChatConversacionResponse> {
    return this.http.patch<ChatConversacionResponse>(
      `${this.apiUrl}/${conversacionId}/cerrar`,
      {}
    );
  }
}
